src/uct/ib/ud/accel/ud_mlx5.c

```diff
@@ -176,79 +189,149 @@ static UCS_CLASS_DEFINE_NEW_FUNC(uct_ud_mlx5_ep_t, uct_ep_t, uct_iface_h,
+                               /* iov data */    const uct_iov_t *iov, size_t iovcnt,
+                               uint32_t packet_flags, uct_completion_t *comp,
+                               unsigned stat_ops_counter, unsigned stat_bytes_counter,
+                               const char *func_name)
 {
-    uct_ud_mlx5_ep_t *ep = ucs_derived_of(tl_ep, uct_ud_mlx5_ep_t);
     uct_ud_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface,
                                                 uct_ud_mlx5_iface_t);
-    size_t ctrl_av_size = uct_ud_mlx5_ep_ctrl_av_size(ep);
-    struct mlx5_wqe_ctrl_seg *ctrl;
+    uct_ud_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_ud_mlx5_ep_t);
     struct mlx5_wqe_inl_data_seg *inl;
-    uct_ud_am_short_hdr_t *am;
-    uct_ud_neth_t *neth;
+    struct mlx5_wqe_ctrl_seg *ctrl;
+    size_t inline_size, wqe_size;
     uct_ud_send_skb_t *skb;
-    size_t wqe_size;
+    ucs_status_t status;
+    uct_ud_neth_t *neth;
+    void *wqe_data;
 
-    /* data a written directly into tx wqe, so it is impossible to use
-     * common ud am code
-     */
-    UCT_CHECK_AM_ID(id);
-    UCT_CHECK_LENGTH(sizeof(uct_ud_neth_t) + sizeof(hdr) + length,
-                    0, iface->super.config.max_inline, "am_short");
+    UCT_CHECK_AM_ID(am_id);
+    UCT_UD_CHECK_ZCOPY_LENGTH(&iface->super, header_size + data_size,
+                              uct_iov_total_length(iov, iovcnt));
+    UCT_CHECK_IOV_SIZE(iovcnt, uct_ud_mlx5_max_am_iov(), func_name);
 
     uct_ud_enter(&iface->super);
 
     skb = uct_ud_ep_get_tx_skb(&iface->super, &ep->super);
     if (!skb) {
-        uct_ud_leave(&iface->super);
-        return UCS_ERR_NO_RESOURCE;
+        status = UCS_ERR_NO_RESOURCE;
+        goto out;
+    }
+
+    wqe_size        = uct_ud_mlx5_ep_get_next_wqe(iface, ep, &ctrl, (void**)&inl);
+    inline_size     = sizeof(*neth) + header_size + data_size;
+    inl->byte_count = htonl(inline_size | MLX5_INLINE_SEG);
+    wqe_size       += sizeof(*inl) + inline_size;
+    skb->len        = inline_size;
+
+    /* set network header */
+    neth              = (void*)(inl + 1);
+    neth->packet_type = (am_id << UCT_UD_PACKET_AM_ID_SHIFT) |
+                        ep->super.dest_ep_id |
+                        packet_flags;
+    uct_ud_neth_init_data(&ep->super, neth);
+    if (!(packet_flags & UCT_UD_PACKET_FLAG_ACK_REQ)) {
+        /* check for ACK_REQ, if not already enabled by packet_flags */
+        neth->packet_type |= uct_ud_ep_req_ack(&ep->super) << UCT_UD_PACKET_ACK_REQ_SHIFT;
     }
 
-    ctrl = iface->tx.wq.curr;
-    /* Set inline segment which has AM id, AM header, and AM payload */
-    inl = uct_ib_mlx5_txwq_wrap_exact(&iface->tx.wq,
-                                      UCS_PTR_BYTE_OFFSET(ctrl, ctrl_av_size));
-    wqe_size = length + sizeof(*am) + sizeof(*neth);
-    inl->byte_count = htonl(wqe_size | MLX5_INLINE_SEG);
+    /* copy inline "header", assume it fits to one BB */
```

Comment:
why need to limit it? 
also should max_zcopy_hdr cap adjusted accordingly?

Reply:
added comment